File: src/api/mcp.ts
import { commands } from "../bindings";
import type {
  AppError,
  McpCatalogSnapshot,
  McpOAuthStart,
  McpServerConfig,
  McpSystemSettings,
  McpToolPolicy,
  McpToolPolicyRequest,
} from "../bindings";

function initialPreviewCatalog(): McpCatalogSnapshot {
  return {
    configPath: "~/Library/Application Support/com.carrot.llm-client/mcp-servers.toml",
    system: {
      controlledLocalTools: true,
      remoteHttp: true,
      secureAuth: true,
      dynamicUpdates: true,
    },
    servers: [
      {
        config: {
          id: "workspace-files",
          label: "Workspace Files",
          enabled: true,
          transport: "stdio",
          executable: "/opt/homebrew/bin/npx",
          arguments: ["-y", "@modelcontextprotocol/server-filesystem", "/Users/lee/Documents"],
          workingDirectory: "/Users/lee/Documents",
          url: null,
          auth: "none",
          oauthClientId: null,
          oauthScopes: [],
          allowedDirectories: ["/Users/lee/Documents"],
          allowNetwork: false,
          toolPolicies: [],
        },
        state: "ready",
        error: null,
        authConfigured: true,
        catalogRevision: "1",
        tools: [
          {
            name: "read_file",
            alias: "mcp_workspace_files_read_file_70e1a9b2",
            title: "Read file",
            description: "Read a text file from the configured workspace.",
            schemaHash: "70e1a9b2",
            readOnlyHint: true,
            enabled: true,
            risk: "read_only",
            idempotent: true,
            reconcile: true,
          },
          {
            name: "list_directory",
            alias: "mcp_workspace_files_list_directory_389f10c4",
            title: "List directory",
            description: "List entries in a workspace directory.",
            schemaHash: "389f10c4",
            readOnlyHint: true,
            enabled: true,
            risk: "read_only",
            idempotent: true,
            reconcile: true,
          },
        ],
      },
    ],
    revision: "1",
  };
}
const previewCatalog = initialPreviewCatalog();

export function resetMcpPreview() {
  const initial = initialPreviewCatalog();
  previewCatalog.configPath = initial.configPath;
  previewCatalog.system = initial.system;
  previewCatalog.revision = initial.revision;
  previewCatalog.servers.splice(0, previewCatalog.servers.length, ...initial.servers);
}

function isTauri() {
  return "__TAURI_INTERNALS__" in window;
}

function resultData<T>(
  result: { status: "ok"; data: T } | { status: "error"; error: AppError },
): T {
  if (result.status === "error") throw new Error(result.error.message);
  return result.data;
}

export async function getMcpCatalog() {
  if (!isTauri()) return structuredClone(previewCatalog);
  return resultData(await commands.mcpCatalogGet());
}

export async function updateMcpSystemSettings(settings: McpSystemSettings) {
  if (!isTauri()) {
    previewCatalog.system = structuredClone(settings);
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpSystemSettingsUpdate({ settings }));
}

export async function createMcpServer(config: McpServerConfig) {
  if (!isTauri()) {
    previewCatalog.servers.push({
      config: structuredClone(config),
      state: "disconnected",
      error: null,
      tools: [],
      authConfigured: config.auth === "none",
      catalogRevision: "0",
    });
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpServerCreate({ config }));
}

export async function installMcpPreset(
  preset: "workspace_filesystem" | "brave_search",
  workspacePath: string | null,
) {
  if (!isTauri()) return structuredClone(previewCatalog);
  return resultData(await commands.mcpPresetInstall({ preset, workspacePath }));
}

export async function updateMcpServer(config: McpServerConfig) {
  if (!isTauri()) {
    const server = previewCatalog.servers.find((item) => item.config.id === config.id);
    if (server) server.config = structuredClone(config);
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpServerUpdate({ config }));
}

export async function deleteMcpServer(serverId: string) {
  if (!isTauri()) {
    previewCatalog.servers.splice(
      0,
      previewCatalog.servers.length,
      ...previewCatalog.servers.filter((item) => item.config.id !== serverId),
    );
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpServerDelete(serverId));
}

export async function connectMcpServer(serverId: string) {
  if (!isTauri()) {
    const server = previewCatalog.servers.find((item) => item.config.id === serverId);
    if (server) server.state = "ready";
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpServerConnect(serverId));
}

export async function disconnectMcpServer(serverId: string) {
  if (!isTauri()) {
    const server = previewCatalog.servers.find((item) => item.config.id === serverId);
    if (server) server.state = "disconnected";
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpServerDisconnect(serverId));
}

export async function setMcpToolPolicy(request: McpToolPolicyRequest) {
  if (!isTauri()) {
    const tool = previewCatalog.servers
      .find((item) => item.config.id === request.serverId)
      ?.tools.find((item) => item.name === request.policy.name);
    if (tool) Object.assign(tool, request.policy);
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpToolPolicySet(request));
}

export async function refreshMcpServer(serverId: string) {
  if (!isTauri()) return structuredClone(previewCatalog);
  return resultData(await commands.mcpServerRefresh(serverId));
}

export async function setMcpAuth(serverId: string, secret: string) {
  if (!isTauri()) {
    const server = previewCatalog.servers.find((item) => item.config.id === serverId);
    if (server) server.authConfigured = true;
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpAuthSet({ serverId, secret }));
}

export async function clearMcpAuth(serverId: string) {
  if (!isTauri()) {
    const server = previewCatalog.servers.find((item) => item.config.id === serverId);
    if (server) server.authConfigured = false;
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpAuthClear(serverId));
}

export async function beginMcpOAuth(serverId: string, redirectUri: string): Promise<McpOAuthStart> {
  if (!isTauri()) {
    return {
      serverId,
      authorizationUrl: "https://example.com/oauth/authorize?state=preview",
    };
  }
  return resultData(await commands.mcpOauthBegin({ serverId, redirectUri }));
}

export async function completeMcpOAuth(serverId: string, callbackUrl: string) {
  if (!isTauri()) {
    const server = previewCatalog.servers.find((item) => item.config.id === serverId);
    if (server) server.authConfigured = true;
    return structuredClone(previewCatalog);
  }
  return resultData(await commands.mcpOauthComplete({ serverId, callbackUrl }));
}

export function policyFromTool(tool: {
  name: string;
  enabled: boolean;
  risk: McpToolPolicy["risk"];
  idempotent: boolean;
  reconcile: boolean;
}): McpToolPolicy {
  return {
    name: tool.name,
    enabled: tool.enabled,
    risk: tool.risk,
    idempotent: tool.idempotent,
    reconcile: tool.reconcile,
  };
}

File: src-tauri/src/commands/mcp.rs
use serde::Deserialize;
use specta::Type;
use tauri::State;

use crate::application::CarrotService;
use crate::domain::mcp::{
    McpCatalogSnapshot, McpOAuthStart, McpPresetKind, McpServerConfig, McpSystemSettings,
    McpToolPolicy,
};
use crate::error::AppError;

#[derive(Debug, Clone, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct McpServerRequest {
    pub config: McpServerConfig,
}

#[derive(Debug, Clone, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct McpPresetInstallRequest {
    pub preset: McpPresetKind,
    pub workspace_path: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct McpSystemSettingsRequest {
    pub settings: McpSystemSettings,
}

#[derive(Debug, Clone, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct McpToolPolicyRequest {
    pub server_id: String,
    pub policy: McpToolPolicy,
}

#[derive(Debug, Clone, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct McpAuthRequest {
    pub server_id: String,
    pub secret: String,
}

#[derive(Debug, Clone, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct McpOAuthBeginRequest {
    pub server_id: String,
    pub redirect_uri: String,
}

#[derive(Debug, Clone, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct McpOAuthCompleteRequest {
    pub server_id: String,
    pub callback_url: String,
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_catalog_get(
    service: State<'_, CarrotService>,
) -> Result<McpCatalogSnapshot, AppError> {
    Ok(service.mcp_snapshot().await)
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_system_settings_update(
    service: State<'_, CarrotService>,
    request: McpSystemSettingsRequest,
) -> Result<McpCatalogSnapshot, AppError> {
    service.update_mcp_system_settings(request.settings).await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_preset_install(
    service: State<'_, CarrotService>,
    request: McpPresetInstallRequest,
) -> Result<McpCatalogSnapshot, AppError> {
    service
        .install_mcp_preset(request.preset, request.workspace_path)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_server_create(
    service: State<'_, CarrotService>,
    request: McpServerRequest,
) -> Result<McpCatalogSnapshot, AppError> {
    service.create_mcp_server(request.config).await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_server_update(
    service: State<'_, CarrotService>,
    request: McpServerRequest,
) -> Result<McpCatalogSnapshot, AppError> {
    service.update_mcp_server(request.config).await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_server_delete(
    service: State<'_, CarrotService>,
    server_id: String,
) -> Result<McpCatalogSnapshot, AppError> {
    service.delete_mcp_server(&server_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_server_connect(
    service: State<'_, CarrotService>,
    server_id: String,
) -> Result<McpCatalogSnapshot, AppError> {
    service.connect_mcp_server(&server_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_server_disconnect(
    service: State<'_, CarrotService>,
    server_id: String,
) -> Result<McpCatalogSnapshot, AppError> {
    service.disconnect_mcp_server(&server_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_tool_policy_set(
    service: State<'_, CarrotService>,
    request: McpToolPolicyRequest,
) -> Result<McpCatalogSnapshot, AppError> {
    service
        .set_mcp_tool_policy(&request.server_id, request.policy)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_auth_set(
    service: State<'_, CarrotService>,
    request: McpAuthRequest,
) -> Result<McpCatalogSnapshot, AppError> {
    service
        .set_mcp_auth(&request.server_id, request.secret)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_auth_clear(
    service: State<'_, CarrotService>,
    server_id: String,
) -> Result<McpCatalogSnapshot, AppError> {
    service.clear_mcp_auth(&server_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_server_refresh(
    service: State<'_, CarrotService>,
    server_id: String,
) -> Result<McpCatalogSnapshot, AppError> {
    service.refresh_mcp_server(&server_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_oauth_begin(
    service: State<'_, CarrotService>,
    request: McpOAuthBeginRequest,
) -> Result<McpOAuthStart, AppError> {
    service
        .begin_mcp_oauth(&request.server_id, request.redirect_uri)
        .await
}

#[tauri::command]
#[specta::specta]
pub async fn mcp_oauth_complete(
    service: State<'_, CarrotService>,
    request: McpOAuthCompleteRequest,
) -> Result<McpCatalogSnapshot, AppError> {
    service
        .complete_mcp_oauth(&request.server_id, request.callback_url)
        .await
}
